### DISCUSSION TITLE
Intended purpose of the URI_COUNT_ODD rule

### DISCUSSION THREAD

**Discussion by trantor** [rspamd/rspamd#3831] [Q&A]
In `rules/misc.lua` we can find the rule below

```
rspamd_config.URI_COUNT_ODD = {
  callback = function (task)
    local ct = task:get_header('Content-Type')
    if (ct and ct:lower():find('^multipart/alternative')) then
      local urls = task:get_urls() or {}
      local nurls = fun.filter(function(url)
        return not url:is_html_displayed()
      end, urls):foldl(function(acc, val) return acc + val:get_count() end, 0)

      if nurls % 2 == 1 then
        return true, 1.0, tostring(nurls)
      end
    end
  end,
  description = 'Odd number of URIs in multipart/alternative message',
  score = 1.0,
  group = 'url',
}

```

The description is... rather descriptive, I might say.
However I can't find any reason why an odd (as is not even) number of URLs might amount to anything suspicious.
The pull request which added the rule back in 2016 and the commit messages are equally lacking in details.

Can anyone provide some more information on this?

**Comment by fatalbanana** [comment]
AFAIU the idea was to signal presence of URLs which are in one of HTML or text parts _only_ rather than both.

**Comment by fa11enangel** [comment]
For most of my "confirmation mails" it is wrong, as I expect the HTML part to habe two Links and text part only one.

HTML Part should have a link with text and optional an URL the user can copy in the browser:

* Please confirm: <a href="https://my.site/confirm?id=...">Confirm now!</a>
* If nothing happens, please click here or copy this URL into your browser: https://my.site/confirm?id=...

As the Text mail can't have HTML, I put only this link there:

* Please click here or copy this URL into your browser: https://my.site/confirm?id=...

This way they are always OD by design, but increase the amount of confirmations significantly. As it is only a small negative factor and I try to do a lot of other things right, this is not a spam/ham thing for me. This should be just an explanation, where this rule might be wrong.
